lib/maths/CLocalOutlierFactors.cc

```diff
@@ -80,7 +80,7 @@ bool computeOutliersNoPartitions(std::size_t numberThreads, core::CDataFrame& fr
 
     auto rowsToPoints = [&points](TRowItr beginRows, TRowItr endRows) {
         for (auto row = beginRows; row != endRows; ++row) {
-            points[row->index()] = TVector{row->data(), row->numberColumns()};
+            new (&points[row->index()]) TVector{row->data(), row->numberColumns()};
```

Comment:
Is it really necessary to use placement new here?  It could cause some subtle bug to be introduced in the future if a future version of one of the classes involved here ever has a non-trivial destructor.

Is there a way of achieving similar efficiency that doesn't involve placement new?  If assigning the whole `TVector` has proved to be highly inefficient for some reason, how about assigning the two members individually?

Reply:
This is the only way you can change the value underlying values referenced by an Eigen::Map: see https://eigen.tuxfamily.org/dox/group__TutorialMapClass.html.

Reply:
I suspect that it is overloaded for this class although haven't looked at the details

Reply:
By way of background, I should add that I want to reference memory in the `CDataFrame` directly, so no copying is wanted. I also want to pre-size the vector for Eigen::Map objects, so I don't need to lock on push_buck. This seems to be the only/designed way of reseating such objects.

Reply:
Based on the docs you could change lines 78-79 from:

```
    CFloatStorage initial;
    TVectorVec points(frame.numberRows(), TVector{&initial, 1});
```

to:

```
    TVectorVec points(frame.numberRows(), TVector{nullptr});
```

That might help avoid any side effects of not calling a destructor of an object overwritten by the placement new after some future change.

Reply:
Yes, this is definitely better and should mean we'll immediately get a fault as soon as an object is used if it wasn't reset rather than unpredictable behaviour. (I'm pretty sure this should be future safe since Eigen::Map is essentially a pointer.)

Reply:
[bfd91e3](https://github.com/elastic/ml-cpp/pull/334/commits/bfd91e33b4f550238e4a395573434748d6a9a49d)